grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,287 @@
+        final ServingStatus status = checkServingStatus(service);
+        if (status == ServingStatus.SERVICE_UNKNOWN) {
+            responseObserver.onError(Status.NOT_FOUND
+                                             .withDescription(String.format(
+                                                     "The service name(=%s) is not registered in this service",
+                                                     service))
+                                             .asRuntimeException());
+            return;
+        }
+        responseObserver.onNext(getHealthCheckResponse(status));
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void watch(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        final HealthCheckResponse response = getHealthCheckResponse(status);
+        responseObserver.onNext(response);
+        watchers.put(responseObserver, Maps.immutableEntry(service, response.getStatus()));
+        ((ServerCallStreamObserver<HealthCheckResponse>) responseObserver).setOnCancelHandler(() -> {
+            watchers.remove(responseObserver);
+        });
+    }
+
+    void serviceAdded(ServiceConfig cfg) {
+        if (server != null) {
+            if (server != cfg.server()) {
+                throw new IllegalStateException("Cannot be added to more than one server");
+            } else {
+                return;
+            }
+        }
+
+        server = cfg.server();
+        server.addListener(new ServerListenerAdapter() {
+            @Override
+            public void serverStarted(Server server) {
+                serverHealth.setHealthy(true);
+            }
+
+            @Override
+            public void serverStopping(Server server) {
+                serverHealth.setHealthy(false);
+            }
+        });
+    }
+
+    @VisibleForTesting
+    ServingStatus checkServingStatus(@Nullable String serverName) {
+        if (!isHealthy()) {
+            return ServingStatus.NOT_SERVING;
+        }
+        if (Strings.isNullOrEmpty(serverName)) {
+            if (grpcServiceHealthCheckers.isEmpty()) {
+                return ServingStatus.SERVING;
+            }
+            if (grpcServiceHealthCheckers.values().stream().allMatch(HealthChecker::isHealthy)) {
+                return ServingStatus.SERVING;
+            } else {
+                return ServingStatus.NOT_SERVING;
+            }
+        } else {
+            if (grpcServiceHealthCheckers.containsKey(serverName)) {
+                if (grpcServiceHealthCheckers.get(serverName).isHealthy()) {
+                    return ServingStatus.SERVING;
+                } else {
+                    return ServingStatus.NOT_SERVING;
+                }
+            } else {
+                return ServingStatus.SERVICE_UNKNOWN;
+            }
+        }
+    }
```

Comment:
nit: How about using early return? 
```suggestion
        if (Strings.isNullOrEmpty(serverName)) {
            if (grpcServiceHealthCheckers.isEmpty()) {
                return ServingStatus.SERVING;
            }
            if (grpcServiceHealthCheckers.values().stream().allMatch(HealthChecker::isHealthy)) {
                return ServingStatus.SERVING;
            }
            return ServingStatus.NOT_SERVING;
        }
        final ListenableHealthChecker listenableHealthChecker = grpcServiceHealthCheckers.get(serverName);
        if (listenableHealthChecker == null) {
            return ServingStatus.SERVICE_UNKNOWN;
        }
        if (listenableHealthChecker.isHealthy()) {
            return ServingStatus.SERVING;
        }
        return ServingStatus.NOT_SERVING;
    }
```
I also assigned `grpcServiceHealthCheckers.get(serverName)` to a variable not to check the map twice. :wink: